src/app/cli/src/cli_validation.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use internal_error::InternalError;
+use opendatafabric::{
+    DatasetName,
+    DatasetPattern,
+    DatasetRef,
+    DatasetRefAny,
+    DatasetRefPattern,
+    DatasetRefRemote,
+    Multihash,
+    RepoName,
+};
+use thiserror::Error;
+use url::Url;
+
+#[derive(Clone)]
+pub struct DatasetPatternValidationRes {
+    pub pattern: DatasetPattern,
+    pub is_dataset_ref: bool,
+}
```

Comment:
Not sure why we need `DatasetPatternValidationRes` type. Is this purely for optimization when instead of a pattern we get a reference without any wildcards?

If so, it might be better for `DatasetRefPattern` type to have `has_wildcards() -> bool` method and `as_dataset_ref() -> Option<DatasetRef>` methods.

This would enable code like:

```rust
fn get_datasets_matching(self, pattern: DatasetRefPattern) -> Stream<DatasetHandle> {
  if !pattern.has_wildcards() {
    // Optimization for when a pattern refers to one dataset
    stream::once(self.get_dataset(pattern.as_dataset_ref().unwrap()))
  } else {
    // <do the expensive matching>
  } 
}
```